torch/distributed/_functional_collectives.py

```diff
@@ -248,6 +248,39 @@ def all_gather_tensor(
     return res
 
 
+def all_gather_tensor_autograd(
+    self: torch.Tensor,
+    gather_dim: int,
+    group: RANK_TYPES,
+    tag: str = "",
+):
+    """
+    Gather tensor data across from all machines and concatenate over ``gather_dim``.
+
+    Note that it currently only supports gather_dim = 0.
+
+    This function is the same as all_gather_tensor but will propagate the
+    backwards gradient across workers.
+
+    See all_gather_tensor for more details on usage.
+    """
+    group_name = _resolve_group_name(group, tag)
+    group_size = c10d._get_group_size_by_name(group_name)
+
+    tensor = torch.ops._c10d_functional_autograd.all_gather_into_tensor(
+        self, group_size, group_name
+    )
+    res = _FromTorchTensor.apply(tensor)
+    # TODO this should be done inside AsyncCollectiveTensor to delay the wait() call
+    if gather_dim != 0:
```

Comment:
Another follow up we should consider (not in this PR):

I think `gather_dim` in allgather_tensor and `scatter_dim` in reduce_scatter_tensor today are loosely defined with regards to the AsyncCollectiveTensor semantic and its behavior are also not well defined from the NCCL allgather/reduce_scatter prospective.

Maybe we should consider removing them in functional collectives (move the relevant logic back to DTensor), to make the collective API more aligned with both c10d and NCCL API. cc @yifuwang